scripts/actions/deserialize-html.js

```diff
@@ -0,0 +1,102 @@
+const unified = require('unified');
+const parse = require('rehype-parse');
+const rehype2remark = require('rehype-remark');
+const stringify = require('remark-stringify');
+const frontmatter = require('remark-frontmatter');
+const remarkMdx = require('remark-mdx');
+const remarkMdxjs = require('remark-mdxjs');
+const handlers = require('./utils/handlers');
+const defaultHandlers = require('hast-util-to-mdast/lib/handlers');
+const heading = require('hast-util-to-mdast/lib/handlers/heading');
+const u = require('unist-builder');
+const { last } = require('lodash');
+
+const component = (h, node) => {
+  if (!node.properties || !node.properties.dataType) {
+    return defaultHandlers[node.tagName](h, node);
+  }
+
+  const { dataType, dataComponent } = node.properties;
+
+  const key =
+    dataType === 'component' ? dataComponent || node.tagName : dataType;
+
+  const handler = handlers[key];
+
+  if (!handler || !handler.deserialize) {
+    throw new Error(
+      `Unable to deserialize node: '${key}'. Please specify a deserializer in 'scripts/actions/utils/handlers.js'`
+    );
+  }
+
+  return handler.deserialize(h, node);
+};
+
+const headingWithCustomId = (h, node) => {
+  const result = heading(h, node);
+  const { id } = node.properties || {};
+
+  if (!id) {
+    return result;
+  }
+
+  const value = `#${id}`;
+  const lastChild = last(result.children);
+  const linkReference = u(
+    'linkReference',
+    {
+      identifier: value,
+      label: value,
+      referenceType: 'shortcut',
+    },
+    [u('text', value)]
+  );
+
+  if (lastChild.type === 'text' && !lastChild.value.match(/\s$/)) {
+    lastChild.value = `${lastChild.value} `;
+  } else {
+    result.children.push(u('text', ' '));
+  }
+
+  result.children.push(linkReference);
+
+  return result;
+};
+
+const processor = unified()
+  .use(parse)
+  .use(rehype2remark, {
+    handlers: {
+      code: component,
+      table: component,
+      thead: component,
+      tbody: component,
+      tr: component,
+      td: component,
+      th: component,
+      span: component,
+      div: component,
+      h1: headingWithCustomId,
+      h2: headingWithCustomId,
+      h3: headingWithCustomId,
+      h4: headingWithCustomId,
+      h5: headingWithCustomId,
+      h6: headingWithCustomId,
```

Comment:
do we usually have headings with custom ids on anything other than an h2 tag?

Reply:
Not sure I can say "usually", but its possible they are there. I'd rather not restrict a content contributor from being able to add a custom ID to an h6, so figured we'd cover all cases.

Reply:
good point!